pkg/networking/session.go

```diff
@@ -0,0 +1,411 @@
+	s := &Session{
+		g:       execution.NewTaskGroup(suppressContextCancellationError),
+		ctx:     sCtx,
+		cancel:  cancel,
+		config:  config,
+		tp:      tp,
+		conn:    conn,
+		bufRead: bufio.NewReader(conn),
+		sendCh:  make(chan *sendPacket, 1), // TODO: Make the size of send channel configurable.
+	}
+
+	attributes := []any{
+		slog.String("namespace", Namespace),
+		slog.String("remote", s.RemoteAddr().String()),
+	}
+	attributes = append(attributes, config.attributes...)
+
+	if config.logger == nil {
+		s.logger = slog.Default().With(attributes...)
+	} else {
+		s.logger = config.logger.With(attributes...)
+	}
+
+	s.g.Run(s.receiveLoop)
+	s.g.Run(s.sendLoop)
+	if s.config.keepAlive {
+		s.g.Run(s.keepaliveLoop)
+	}
+
+	return s, nil
+}
+
+func (s *Session) String() string {
+	return fmt.Sprintf("Session{local=%s,remote=%s}", s.LocalAddr(), s.RemoteAddr())
+}
+
+// LocalAddr returns the local network address.
+func (s *Session) LocalAddr() net.Addr {
+	if a, ok := s.conn.(addressable); ok {
+		return a.LocalAddr()
+	}
+	return &sessionAddress{addr: "local"}
+}
+
+// RemoteAddr returns the remote network address.
+func (s *Session) RemoteAddr() net.Addr {
+	if a, ok := s.conn.(addressable); ok {
+		return a.RemoteAddr()
+	}
+	return &sessionAddress{addr: "remote"}
+}
+
+// Close is used to close the session. It is safe to call Close multiple times from different goroutines,
+// subsequent calls do nothing.
+func (s *Session) Close() error {
+	s.shutdownLock.Lock()
+	defer s.shutdownLock.Unlock()
+
+	if s.shutdown {
+		return nil // Fast path - session already closed.
+	}
+	s.shutdown = true
+
+	s.logger.Debug("Closing session")
+	clErr := s.conn.Close() // Close the underlying connection.
+	if clErr != nil {
+		s.logger.Warn("Failed to close underlying connection", "error", clErr)
+	}
+	s.logger.Debug("Underlying connection closed")
+
+	s.cancel() // Cancel the underlying context to interrupt the loops.
+
+	s.logger.Debug("Waiting for loops to finish")
+	err := s.g.Wait() // Wait for loops to finish.
+
+	err = errors.Join(err, clErr) // Combine loops finalization errors with connection close error.
+
+	s.logger.Debug("Session closed", "error", err)
+	return err
+}
+
+// Write is used to write to the session. It is safe to call Write and/or Close concurrently.
+func (s *Session) Write(msg []byte) (int, error) {
+	s.sendLock.Lock()
+	defer s.sendLock.Unlock()
+
+	if err := s.waitForSend(msg); err != nil {
+		return 0, err
+	}
+
+	return len(msg), nil
+}
+
+// waitForSend waits to send a data, checking for a potential context cancellation.
+func (s *Session) waitForSend(data []byte) error {
+	// Channel to receive an error from sendLoop goroutine.
+	// We are not closing this channel, it will be GCed when the session is closed.
+	errCh := make(chan error, 1)
+
+	timer := s.tp.Get()
+	timer.Reset(s.config.connectionWriteTimeout)
+	defer s.tp.Put(timer)
+
+	s.logger.Debug("Sending data", "data", base64.StdEncoding.EncodeToString(data))
```

Comment:
Doesn't work this way. B64Bytes.String() doesn't called in this case:
```go
s.logger.Debug("Message", "data", B64Bytes(data))
```

Reply:
Fixed by wrapping heavy logs in level checks.

Reply:
> Doesn't work this way. B64Bytes.String() doesn't called in this case:
> 
> ```go
> s.logger.Debug("Message", "data", B64Bytes(data))
> ```


@alexeykiselev This will work
```go
type b64Bytes struct{ data []byte }

func (b b64Bytes) String() string { return base64.StdEncoding.EncodeToString(b.data) }
```
`type B64Bytes []byte` is just a wrapper for bytes. For `go` it is still bytes at the internal level, though the wrapper above is not.

Reply:
> Fixed by wrapping heavy logs in level checks.

What if `slog.Handler` is `discardHandler` with `DEBUG` log level?

Reply:
> Fixed by wrapping heavy logs in level checks.

UPD: `discardHandler` returns `false` for every `Enabled` cal with any arguments. So the solution is good.